### DISCUSSION TITLE
Specify namespace to where fluxv2 should create resources

### DISCUSSION THREAD

**Discussion by dimakyriakov** [fluxcd/flux2#2909] [Q&A]
Currently my resources that are going to be created - don't have "namespace" specified in metadata, because I want to create them depending on the namespace that I will specify while fluxv2 bootstrap.
I already tried to set "targetNamespace: <name>" in Flux Kustomization file in this way:
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- gotk-components.yaml
- gotk-sync.yaml
spec:
  targetNamespace: <namespace>
```
And I am getting an error:
`✗ json: unknown field "spec"`

Is there a way to set my namespace name during bootstrap or directly in flux files?
As result I expect that flux will do `kubectl -n <namespace name> apply -f <resource>`     
(I know that maybe fluxv2 not executing exactly that command I gave, but I hope you understood what I am saying)

**Comment by hiddeco** [comment]
You seem to be mixing a Flux Kustomization with a Kustomize Kustomization, see https://fluxcd.io/docs/faq/#are-there-two-kustomization-types.

**Reply by dimakyriakov** [reply]
By default, it creates v1beta1, as I understood I need v1beta2 to use "targetNamespace" option.
Do you have some advice on how to make it work? Is it possible to choose such apiversion by executing "fluxv2 bootstrap"?

**Reply by hiddeco** [reply]
They are different files, one is a `kustomization.yaml` file containing an `apiVersion` with the `kustomization.kustomize.config.k8s.io` domain. The other is a separate file containing a Kubernetes resource which can be applied using `kubectl apply -f` with an `apiVersion` with the `kustomization.kustomize.toolkit.fluxcd.io` domain (which is different from the `kustomization.yaml` file).

You can find the specification for the latter object here: https://fluxcd.io/docs/components/kustomize/kustomization/

**Reply by dimakyriakov** [reply]
I still getting "namespace not specified, error: the server could not find the requested resource" error, despite specified targetNamespace: <name> value.
![image](https://user-images.githubusercontent.com/96116626/179494203-5e507078-b951-46e6-952f-8814ef00ea85.png)
![image](https://user-images.githubusercontent.com/96116626/179494177-7056b833-bbb9-4cac-8c20-9931ef67ff68.png)

**Reply by hiddeco** [reply]
The controller does not create the resource for you, but it is expected to exist and have been created by another Kustomization.

**Reply by dimakyriakov** [reply]
sorry, I think we miss understood each other. I want flux to apply files where I didn't specify namespace value. Is it possible for fluxv2 to create resources in a specific namespace?

**Reply by kingdonb** [reply]
There are these two methods to achieve a "namespace override" and you appear to have them conflated, as Hidde was explaining.

In `kustomization.kustomize.config.k8s.io` there is `.namespace` at the top level, which you can set to override a blank or other namespace in the applied resources.

In Flux's `kustomization.kustomize.toolkit.fluxcd.io` there is `spec.targetNamespace` which it seems you have found. I suspect there is an order of operations problem now. You need to be sure that `spec.targetNamespace` is set inside the cluster before `Secret/aws-secret` is being created, because Flux is going to use the Kustomization that is already in the cluster with its settings to do the apply, not the Kustomization that is in the file (which hasn't been applied yet, with its setting to correct the namespace.)

Since the Kustomize CLI overlay `kustomization.yaml` is not a Kubernetes resource it does not have this issue.

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: <namespace>
resources:
...
```

... but you'll need to create a separate `kustomization.yaml` for this to work, as the one you mentioned at the top of the post is used to apply Flux, which will be expected to remain in the `flux-system` namespace. Hope this helps!
